### ISSUE TITLE
aws-vault server should require `Host: 169.254.169.254` header

### ISSUE THREAD

**Issue by philsnow** [99designs/aws-vault#578]
tl;dr:  With the aws-vault EC2 metadata service running, it's possible for an attacker controlling any web page the victim visits with javascript enabled to mount a [DNS rebinding attack](https://en.wikipedia.org/wiki/DNS_rebinding) to grab short-lived credentials from the aws-vault server.

If it were to require an explicit `Host: 169.254.169.254` header, that attack would be mitigated.

## Setup

Run `aws-vault exec --server <profile>`.  It creates an alias IP 169.254.169.254 on the loopback interface, starts a proxy bound to 169.254.169.254:80 that just proxies to http://127.0.0.1:9099/, and starts an http server listening on localhost:9099 that exposes short-lived credentials, similar to how the EC2 metadata service does.

Put the following HTML somewhere where you can serve it with plain http on port 80:

```html
<html>
    <body>
	<script type="text/javascript">
	 var xhr = new XMLHttpRequest();
	 xhr.onload = function() {
	     console.log(xhr.responseText);
	 }
	 xhr.open("GET", "http://www.example.org/latest/meta-data/iam/security-credentials/local-credentials");
	</script>
	<input type="button" onclick="xhr.send(null);" value="click me">
    </body>
</html>
```

Let's say you're serving that on http://1.2.3.4:80/index.html.  Add a line to your /etc/hosts `1.2.3.4 www.example.org`, flush all your DNS caches (`sudo killall -HUP mDNSResponder` on mac, "Clear host cache" button on chrome://net-internals#dns, whatever).  Load that page in Chrome by going to http://www.example.org/index.html .

Edit your /etc/hosts line to be `169.254.169.254 www.example.org`, flush all your DNS caches, then click the "click me" button on index.html.

### Result

Chrome re-resolves the name www.example.org to 169.254.169.254, allows the XHR to that IP, and logs the short-lived credential to the console.  In a real attack, the attacker would exfiltrate it.

Although I used a button in the above PoC, nothing about this attack requires the victim to click or interact with anything;  as long as they load a malicious page and have javascript enabled it can work.  A typical DNS rebinding attack uses DNS records with short TTL (and a DNS server that actively tries to work around things like Chrome's in-browser DNS cache).

## Mitigation

In the above example, Chrome sends a header `Host: www.example.org`.  If aws-vault's server required a header `Host: 169.254.169.254`, that would mitigate this attack.  Chrome will block  XHR to http://169.254.169.254/... from www.example.org because aws-vault doesn't send any CORS headers that would allow that access.

### Possible downsides

While botocore explicitly uses the literal http://169.254.169.254:80 during the `iam-role` step of the credentials chain provider, I haven't looked at all the SDKs to make sure they use that IP; it's possible they some DNS name that happens to resolve to 169.254.169.254 inside AWS's split-horizon DNS.

If that turns out to be the case, we could control this behavior (of requiring the header) with a command line switch that is on by default, and let the user decide whether to disable it (`--no-require-host-header` or something).

**Comment by philsnow** [comment]
Another mitigation could well be removing support for the EC2 metadata service in favor of the ECS metadata service (as soon as all the issues with the latter are worked out) because it uses URLs with UUIDs in them that are infeasible for an attacker to guess, rather than a single well-known URL.

A downside of doing this would be then you have to communicate the `AWS_CONTAINER_CREDENTIALS_RELATIVE_URI` environment variable to anything that needs to consume the service, whereas if you use the EC2 metadata service, you just let botocore try to use the default credentials chain provider and it succeeds.  There would still be value, though, since this would allow long-running processes to keep using short-lived credentials (with MFA), while also making a DNS rebinding attack infeasible.

**Comment by philsnow** [comment]
I'd be happy to take a stab at a PR to require the Host header, if y'all think it's the right thing to do.

**Comment by frezbo** [comment]
@philsnow Most of the SDK's have the `169.254.169.254` IP hardcoded, I have't seen one using a DNS name. So I assume it's safe to have a `Host` header. I will let other folks chime in.

**Comment by mtibben** [comment]
Nice find @philsnow, thanks for the report. I have a PR incoming for master, we might need to look at a fix for v4 and v5

**Comment by mtibben** [comment]
- v5 fix 38ffc2ab42308e434022b5d699fff95446e8856e
- v4 fix 8a75233e34974c3fd0f8cf25e9b30127e5257603

**Comment by mtibben** [comment]
@philsnow if you're able to verify the fix at #580 or https://github.com/99designs/aws-vault/releases/tag/v5.4.4-rc1 that would be much appreciated

**Comment by mtibben** [comment]
I've verified AWS CLI works with [v5.4.4-rc1](https://github.com/99designs/aws-vault/releases/tag/v5.4.4-rc1). I'd be interested if anyone can test against the various SDKs out there using `aws-vault exec --server`.

**Comment by philsnow** [comment]
![fix confirmation screenshot](https://cdn.zappy.app/6e0dbb0488ad37bb4704a8f925767f66.png)

confirmed, fix works, thanks for the super quick turnaround!

**Comment by pda** [comment]
Great work @philsnow & @mtibben :ok_hand:
